[JCLOUDS-1257] Initial support ManagedDataDisks

andreaturli opened the pull request:
- add DiskApi + tests
- add ImageApi + tests
- add TemplateOptions.dataDisks
- use dataDisks in adapter

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/functions/VMImageToImage.java:
```diff
@@ -71,8 +71,8 @@ public static String encodeFieldsToUniqueIdCustom(VMImage imageReference) {
-      String[] fields = checkNotNull(id, "id").split("/");
       VMImage vmImage;
+      String[] fields = checkNotNull(id, "id").split("/");
       boolean custom = fields.length == 5;
```
Change this to a public static function that can be called from the adapter, instead of duplicating this logic there.

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/options/AzureTemplateOptions.java:
```diff
@@ -33,6 +36,7 @@
    private String blob;
    private AvailabilitySet availabilitySet;
    private String availabilitySetName;
+   protected List<Map<String, String>> dataDisks = ImmutableList.of();
```
Better create an object to carry the information of the data disks? A `Map` looks quite inconvenient and too error-prone, and also harder to document and use by end users.

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/options/AzureTemplateOptions.java:
```diff
@@ -100,44 +117,54 @@ public void copyTo(TemplateOptions to) {
          eTo.blob(blob);
          eTo.availabilitySet(availabilitySet);
          eTo.availabilitySet(availabilitySetName);
+         if (!dataDisks.isEmpty()) {
```
Redundant check. We allow passing an empty list.

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/options/AzureTemplateOptions.java:
```diff
@@ -181,5 +208,18 @@ public static AzureTemplateOptions availabilitySet(String availabilitySetName) {
+      /**
+       * @see AzureTemplateOptions#dataDisks
+       */
+      public static AzureTemplateOptions blockDevices(Map<String, String>... dataDisks) {
```
Why not naming this `dataDisks` if that is the naming used in Azure?

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/strategy/CreateResourceGroupThenCreateNodes.java:
```diff
@@ -136,8 +128,8 @@ protected CreateResourceGroupThenCreateNodes(
-      StorageService storageService = getOrCreateStorageService(group, azureGroupName, location, template.getImage());
-      options.blob(storageService.storageServiceProperties().primaryEndpoints().get("blob"));
+      //StorageService storageService = getOrCreateStorageService(group, azureGroupName, location, template.getImage());
+      //options.blob(storageService.storageServiceProperties().primaryEndpoints().get("blob"));
```
Remove all commented code.

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/CreationData.java:
```diff
@@ -0,0 +1,34 @@
+public abstract class CreationData {
+
+   @Nullable
+   public abstract String createOption();
```
Is there an enum we could use here?

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/DataDisk.java:
```diff
@@ -16,47 +16,140 @@
+               return item;
+            }
+         }
+         return null;
```
Better throw an IAE. What is the purpose of returning `null`?

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/DataDisk.java:
```diff
@@ -16,47 +16,140 @@
+               return item;
+            }
+         }
+         return null;
```
Better throw an IAE. What is the purpose of returning `null`?

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/Disk.java:
```diff
@@ -0,0 +1,97 @@
+      public abstract Builder type(String type);
+      public abstract Builder sku(SKU sku);
+      public abstract Builder properties(DiskProperties properties);
+      public abstract Builder tags(Map<String, String>  tags);
```
Make sure this map is immutable following [this](https://github.com/jclouds/jclouds-labs/blob/master/azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/AvailabilitySet.java#L82-L90) pattern.

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/domain/Image.java:
```diff
@@ -0,0 +1,63 @@
+   public abstract static class Builder {
+      public abstract Builder location(String location);
+      public abstract Builder properties(ImageProperties properties);
+      public abstract Builder tags(Map<String, String>  tags);
```
Make immutable.

nacx commented on azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/DiskApiLiveTest.java:
```diff
@@ -0,0 +1,99 @@
+   @Test(dependsOnMethods = "createDisk")
+   public void listDisks() {
+      List<Disk> dataDisks = api().list();
+      assertTrue(dataDisks.size() > 0);
```
Verify that the created disk is there. We don't care about others? Just in case the RG is not empty.

nacx commented on azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/DiskApiLiveTest.java:
```diff
@@ -0,0 +1,99 @@
+            return waitForState(name, state);
+         }
+      }, 60 * 4 * 1000).apply(diskName);
+   }
```
Better change these two methods to use the already available predicate to poll Provisionable instances:

```java
 private boolean waitUntilAvailable(String name) {
   return resourceAvailable.apply(new Supplier<Provisionable>() {
      @Override public Provisionable get() {
           Disk disk api().get(name);
           return disk == null ? null : disk.properties();
      }
   });
}
```

nacx commented on azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/ImageApiLiveTest.java:
```diff
@@ -0,0 +1,108 @@
+   }
+
+   @Test(dependsOnMethods = "deleteImageResourceDoesNotExist")
+   public void CreateVirtualMachineImageFromExistingVM() {
```
[minor] start lowercase

nacx commented on azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/ImageApiLiveTest.java:
```diff
@@ -0,0 +1,108 @@
+      assertNull(api().delete(imageName));
+   }
+
+   @Test(dependsOnMethods = "deleteImageResourceDoesNotExist")
```
Remove this test dependency. Make the delete unexisting try to delete a randomly generated name, but we don't want the tests to be skipped because of that method failing.

nacx commented on azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/DiskApiLiveTest.java:
```diff
@@ -0,0 +1,99 @@
+      assertNull(api().delete(diskName));
+   }
+
+   @Test(dependsOnMethods = "deleteDiskResourceDoesNotExist")
```
Remove this test dependency. Make the delete unexisting try to delete a randomly generated name, but we don't want the tests to be skipped because of that method failing.

nacx commented on azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/ImageApiLiveTest.java:
```diff
@@ -0,0 +1,108 @@
+            return waitForState(name, state);
+         }
+      }, 60 * 4 * 1000).apply(imageName);
+   }
```
Same comment about using the already existing predicate to poll provisionables.

andreaturli commented:
Thanks @nacx for your review, the impl was a bit raw but I wanted to open the PR to start the discussion in the open, especially with @danielestevez

As for the plan, yes it is exactly what I had in mind!

danielestevez commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/features/ImageApi.java:
```diff
@@ -0,0 +1,81 @@
+   @Fallback(EmptyListOnNotFoundOr404.class)
+   List<Image> list();
+
+   @Named("image:create_or_update")
```
I'm following this https://docs.microsoft.com/en-us/rest/api/manageddisks/images/images-create and it doesn't seem that you can update an image with this API, only update. Have you tried it?
i checked at least you can't change 'sourceVirtualMachine' to an existing Image and name can't be changed either

andreaturli replied:
I will test it asap

danielestevez commented on azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/ImageApiLiveTest.java:
```diff
@@ -0,0 +1,108 @@
+      assertTrue(id.equals(image.properties().sourceVirtualMachine().id()));
+   }
+
+   @Test(dependsOnMethods = "CreateVirtualMachineImageFromExistingVM")
```
As i commented in devlist, we should also try to add thests for creating image from existing disk, blob or snapshot. Maybe later  but good to have TODOs for it here

nacx commented:
Ok, there is a version of the ImageExtension implementation that uses the new Image API working [here](https://github.com/danielestevez/jclouds-labs/tree/fixGeneralizeImage). I'd pretty much like to progress on this PR (happy to help!) so that branch can be rebased on top of this one. This way we'll be able to cleanly merge this first, then the ImageExtension one.

andreaturli commented:
sorry @nacx and @danielestevez for long radio silence, I'm having another look at this.
I've addressed most of the comments and rebased, let's see if the builder is happier.

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/options/AzureTemplateOptions.java:
```diff
@@ -76,12 +83,50 @@ public AzureTemplateOptions availabilitySet(String availabilitySetName) {
       return this;
    }
 
+   public AzureTemplateOptions dataDisks(List<Map<String, String>> dataDisksAsMap) {
```
What is the purpose of keeping the Map version?

andreaturli replied:
I think it'd simplify the consumption from downstream projects like Apache Brooklyn. wdyt?

nacx replied:
The logic to build the DataDisk object from the map is trivial and does not seem to make sense in jclouds. Also, you are limiting it to how you build the object in the options method. It is more flexible to let users build the DataDisk object. In the end, if they know which fields in the map they need, they already know which fields of the DataDisk object to fill, so I don't see the point here.

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/options/AzureTemplateOptions.java:
```diff
@@ -100,44 +145,52 @@ public void copyTo(TemplateOptions to) {
+   }
+
+   @Override
+   public String toString() {
```
Better override the `string()` method instead. Otherwise you won't get the fields from the superclass.

andreaturli replied:
I've implemented dozen of `*TemplateOptions` and I've never noticed `string()` :) 
Thanks!

nacx commented on azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/AzureComputeServiceAdapter.java:
Better use the enum? `CreateOptions.FROM_IMAGE.toString()`

andreaturli commented:
cool @nacx, merging now

andreaturli commented:
pushed to [master](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/51ddc26f)

andreaturli commented:
ported to [2.0.x](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/f4771185)
